### DISCUSSION TITLE
Introducing a new contextual keyword "sub" for interfaces

### DISCUSSION THREAD

**Discussion by nczsl** [dotnet/csharplang#8037] [Language Ideas]
/*
In C#, interfaces are used to constrain their implementing classes. Currently, there is no way to constrain non-public members of an implementing class. As of C# 12, interfaces now support default implementations where the default-implemented members can only be accessed via explicit interface invocation, which helps avoid confusion in cases of multiple interface inheritance. This has indeed introduced more convenience and ease with interface implementation and auto-implementation in C#. While I strongly support this feature, I propose a small improvement: introducing a contextual keyword similar to "value" in properties, called "sub". Contextual keywords lose their special meaning outside their context, thus preventing confusion and namespace pollution, making it a safer technical choice. Here's a usage example:
*/
```csharp
public interface IA
{
  // Requires that an implementing class of IA must have a member named "y" of type int
  int X => sub.y;
  // Another form of the example
  string[] Y
  {
    get => sub.x;
    set => sub.x = value;
  }
}
/*
One key benefit of this proposal is that it would allow implementing classes to avoid writing seemingly redundant property implementations, as shown in the following example:
*/

public interface IA<T>
{
  T Source { get; set; }
}

public class A<T> : IA<T>
{
  public T Source { get; set; }
}

// With the 'sub' keyword, it could look like this:
public interface IA<T>
{
  T Source => sub.source;
  T Source2
  {
    get => sub.source2;
    set => sub.source2.value = value;
  }
  IA<T> _A { get; }
  // ...
}

public class A<T> : IA<T>
{
  private T source; // If not present, indicates IA<T> is not fully implemented
  protected T source2; // If not present, indicates IA<T> is not fully implemented
  public IA<T> _A { get; private set; }
}

// Usage:
//...
void m()
{
  var a = new A<int>();
  var x1 = a._A.Source; // ok
  a._A.Source2 = 33; // ok
}
//...
```

**Comment by HaloFour** [comment]
I don't understand, what defines this `sub` member as being a part of the interface?  There's no declaration of that contract, it only seems to be enforced by virtue of implicitly existing.  That seems very brittle to me.  Also, it seems to give direct access to private state within the implementing class which breaks encapsulation.

**Comment by nczsl** [comment]
/*
An additional point to clarify is that, to prevent complexity and potential errors, it should be specified that within the scope of a property, the member designated by 'sub' must be unique. Multiple designations would be considered an error. An example follows:
*/
```csharp
public interface IA
{
  int Source
  {
    get
    {
      //...
      var x = sub.source + 10; // ok
      return 2 * x; // ok
      // var x2 = sub.source2 + 5; // Error: There is already a 'source' defined within this scope.
      // return sub.source == 0; // Error: An expression involving 'sub' is inconsistent with the property's return type.
      //...
    }
  }
}
```

**Reply by HaloFour** [reply]
Why?  What if the member would need multiple pieces of state?

**Comment by nczsl** [comment]
Further Explanation
After incorporating the safety measures mentioned earlier, the sub contextual keyword becomes more viable, enhancing the functionality of interface default implementations. Consider the following examples:
```csharp
public interface IA
{
    int Source
    {
        get => sub.source;
        private set => sub.source = value;
    }
}

public interface IB
{
    int Source
    {
        get => sub.source2;
        private set => sub.source2 = value;
    }
}

// With the use of 'sub', there will be no conflict when multiple interfaces have properties with the same name during inheritance.
// According to existing C# rules, interface default implementation members can only be accessed through explicit interface member //implementation.

public class X : IA, IB
{
    int source;
    int source2;
}

// Furthermore, the introduction of 'sub' allows setting private setters, which adds meaningful constraints. Implementing classes could privately set these properties through the interface, maintaining encapsulation and consistency.
```

**Comment by iam3yal** [comment]
@nczsl You should edit your posts and format them with proper markdown code blocks.

**Reply by nczsl** [reply]
Ok, I've modified it to markdown format ！

**Comment by HaloFour** [comment]
I'd suggest that if there is to be a shorthand for an interface to declare state that needs to be provided by an implementing class that it be done so more explicitly.  Having that state declared implicitly by virtue of the implementation details of a default method would be quite brittle, not to mention there's nothing about it that would explicitly declare the type of said state.  There's no way for the compiler to know what the type of the expression `sub.foo` is, especially when `foo` does not exist.

I personally don't see what the issue is with the current mechanism for supporting this, namely that the interface declare a `protected` property member which the implementing class must provide.  That gives the same functionality through an explicitly defined contract and proper encapsulation, as the default member can only invoke the member accessible to it through the interface and cannot reach into any private fields of the implementing class.  Sure, it's more verbose, but it's also much more clear and much more explicit.

**Comment by nczsl** [comment]
thanks all. this topic need close, because it not my real need. so i publish a new discussions : https://github.com/dotnet/csharplang/discussions/8040
